src/mbgl/text/glyph_pbf.cpp

```diff
@@ -23,68 +21,96 @@ std::vector<Glyph> parseGlyphPBF(const GlyphRange& glyphRange, const std::string
+                    case 5: // left
+                        glyph.metrics.left = glyph_pbf.get_sint32();
+                        hasLeft = true;
+                        break;
+                    case 6: // top
+                        glyph.metrics.top = glyph_pbf.get_sint32();
+                        hasTop = true;
+                        break;
+                    case 7: // advance
+                        glyph.metrics.advance = glyph_pbf.get_uint32();
+                        hasAdvance = true;
+                        break;
+                    default:
+                        glyph_pbf.skip();
+                        break;
                 }
             }
 
             // Only treat this glyph as a correct glyph if it has all required fields. It also
             // needs to satisfy a few metrics conditions that ensure that the glyph isn't bogus.
             // All other glyphs are malformed.  We're also discarding all glyphs that are outside
             // the expected glyph range.
-            if (!hasID || !hasWidth || !hasHeight || !hasLeft || !hasTop || !hasAdvance ||
-                glyph.metrics.width >= 256 || glyph.metrics.height >= 256 ||
-                glyph.metrics.left < -128 || glyph.metrics.left >= 128 ||
-                glyph.metrics.top < -128 || glyph.metrics.top >= 128 ||
-                glyph.metrics.advance >= 256 ||
+            if (!hasID || !hasWidth || !hasHeight || !hasLeft || !hasTop || !hasAdvance || glyph.metrics.width >= 256 ||
+                glyph.metrics.height >= 256 || glyph.metrics.left < -128 || glyph.metrics.left >= 128 ||
+                glyph.metrics.top < -128 || glyph.metrics.top >= 128 || glyph.metrics.advance >= 256 ||
                 glyph.id < glyphRange.first || glyph.id > glyphRange.second) {
-                continue;
+                return;
             }
 
             // If the area of width/height is non-zero, we need to adjust the expected size
             // with the implicit border size, otherwise we expect there to be no bitmap at all.
             if (glyph.metrics.width && glyph.metrics.height) {
-                const Size size {
-                    glyph.metrics.width + 2 * Glyph::borderSize,
-                    glyph.metrics.height + 2 * Glyph::borderSize
-                };
+                const Size size{glyph.metrics.width + 2 * Glyph::borderSize,
+                                glyph.metrics.height + 2 * Glyph::borderSize};
 
                 if (size.area() != glyphData.size()) {
-                    continue;
+                    return;
                 }
 
                 glyph.bitmap = AlphaImage(size, reinterpret_cast<const uint8_t*>(glyphData.data()), glyphData.size());
             }
 
             result.push_back(std::move(glyph));
+        };
+
+        double ascender{0.0}, descender{0.0};
+        uint16_t count{0};
+        auto fontstack_pbf = glyphs_pbf.get_message();
+        while (fontstack_pbf.next()) {
```

Comment:
could we merge this logic into the existing `while` loop rather than create a new one?

Reply:
The nested while loop is necessary because it reveals different level of the data inside the structure that defined in the proto definition file 
```
// Stores a glyph with metrics and optional SDF bitmap information.
message glyph {
    required uint32 id = 1;

    // A signed distance field of a glyph with buffer documented in metadata.
    optional bytes bitmap = 2;

    // Glyph metrics.
    required uint32 width = 3;
    required uint32 height = 4;
    required sint32 left = 5;
    required sint32 top = 6;
    required uint32 advance = 7;
}

// Stores a face with glyphs and metadata.
message fontstack {
    required string name = 1;
    required string range = 2;
    repeated glyph glyphs = 3;

    optional sint32 ascender = 4;
    optional sint32 descender = 5;
}

message glyphs {
    repeated fontstack stacks = 1;

    extensions 16 to 8191;
}
```